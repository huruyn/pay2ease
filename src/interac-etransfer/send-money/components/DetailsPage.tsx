import { Dispatch, SetStateAction } from 'react';
import {
  Row,
  Col,
  Form,
  FormControl,
  Button,
  Spinner,
} from 'react-bootstrap';
import { Link } from 'react-router-dom';
import { Account, Contact, User } from '../../../api';
import { TransferMainDetails } from '../SendMoney';
import { formatContactName } from '../../../Helpers';

interface DetailsPageProps {
  selectedContact : number;
  setContactToSend: Dispatch<SetStateAction<number>>;
  mainInfo: TransferMainDetails;
  setMainInfo: Dispatch<SetStateAction<TransferMainDetails>>;
  accounts: Account[] | null;
  contacts: Contact[] | null;
  selectedAccount: number;
  setSelectedAccount: Dispatch<SetStateAction<number>>;
  handleSecurity: any;
  isProcessing: boolean;
  user: User | undefined;
}
const DetailsPage = ({
  selectedContact = 0,
  setContactToSend,
  mainInfo,
  setMainInfo,
  accounts,
  contacts,
  selectedAccount,
  setSelectedAccount,
  user,
  handleSecurity,
  isProcessing,
}: DetailsPageProps): JSX.Element => {
  const getUserFullName = () => (user && user.firstName && user.lastName ? `${user.firstName} ${user.lastName}` : '');
  const getUserEmail = () => (user && user?.email);

  const handleContactChange = (evt: any) => {
    setContactToSend(Number(evt.target.value));
  };

  const handleAccountChange = (evt: any) => {
    setSelectedAccount(Number(evt.target.value));
  };

  const getEmail = (id: number) => {
    const contact = id ? contacts?.find((el: Contact) => el.id === id) : 'No email';
    return (contact as Contact)?.email || 'No email';
  };

  const getSelectedContactItem = () => contacts?.find((el: Contact) => el.id === selectedContact);

  return (
    <>
      <h4>Your Interac e-Transfer Details</h4>
      <ul>
        <li>
          Fees apply to Interac e-Transfer transactions.
          {' '}
          <a href="/" className="text-black">
            Learn more about Interac e-Transfer fees
          </a>
        </li>
        <li>
          There are limits on how much money you can send in a day, a week and a
          month.
          {' '}
          <a href="/" className="text-black">
            Learn more about Interac e-Transfer limits
          </a>
        </li>
        <li>
          Note: Sending money outside of Canada? Use the
          {' '}
          <a href="/" className="text-black">
            ZippyX Global Money Transfer Services
          </a>
        </li>
      </ul>
      <Row className="mt-4">
        <Col md={4}>
          <Form.Label>Send Money To:</Form.Label>
        </Col>
        <Col md={8}>
          <Form.Select
            className="send-account-select"
            onChange={handleContactChange}
            value={selectedContact}
          >
            <option>Select</option>
            {contacts?.map((item: Contact) => (
              <option key={item.id} value={item.id}>
                {formatContactName(item.firstName, item.lastName)}
              </option>
            ))}
          </Form.Select>
          <div className="contact-details">
            <div>
              { `Email: ${getEmail(selectedContact)}` }
            </div>
            {contacts && user && (
            <div className="contact-details-controls">
              <Link
                to={`/interac-etransfer/contact-list/edit/${selectedContact}`}
                className={`contact-details-links noselect ${!selectedContact && 'disabled-link'}`}
                state={{ item: getSelectedContactItem() }}
              >
                Edit Contact
              </Link>
              <Link
                to="/interac-etransfer/contact-list/edit"
                className="contact-details-links noselect"
                state={{ item: getSelectedContactItem() }}
              >
                New Contact
              </Link>
            </div>
            )}
          </div>
        </Col>
      </Row>
      <Row className="mt-2">
        <Col md={4}>
          <Form.Label>Amount:</Form.Label>
        </Col>
        <Col md={8}>
          <FormControl
            value={mainInfo?.amount}
            type="number"
            step=".01"
            onChange={(evt) => setMainInfo({ ...mainInfo, amount: Number(evt.target.value) })}
          />
          <Form.Label>
            The maximum amount you can send in each transfer is $3,000.
          </Form.Label>
        </Col>
      </Row>
      <Row className="mt-2">
        <Col md={4}>
          <Form.Label>From Account:</Form.Label>
        </Col>
        <Col md={8}>
          <Form.Select
            className="from-account-info"
            onChange={(evt) => handleAccountChange(evt)}
            value={selectedAccount}
          >
            <option>Select</option>
            {
              accounts?.map((account:Account) => (
                <option key={account.name} value={account.id}>{account.name}</option>
              ))
            }
          </Form.Select>
        </Col>
      </Row>
      <Row className="mt-2">
        <Col md={4}>
          <Form.Label>Transfer Method:</Form.Label>
        </Col>
        <Col md={8}>
          <Form.Select
            className="transfer-method"
            onChange={(evt) => setMainInfo({ ...mainInfo, transferMethod: evt.target.value })}
            value={mainInfo?.transferMethod}
          >
            <option>Email</option>
          </Form.Select>
        </Col>
      </Row>
      <Row className="mt-2">
        <Col md={4}>
          <Form.Label>Message (Optional):</Form.Label>
        </Col>
        <Col md={8}>
          <Form.Control
            as="textarea"
            rows={3}
            onChange={(evt) => setMainInfo({ ...mainInfo, message: evt.target.value })}
            value={mainInfo?.message}
          />
          <p>
            Do not provide the security question, any part of the security
            answer or any confidential information in your message to the
            contact. This message will be viewable on the Interac e-Transfer
            Status page.
          </p>
        </Col>
      </Row>
      <hr style={{ height: '1px', width: '100%' }} />
      <Row className="mt-2">
        <Col md={12}>
          <p>
            This is how you will appear in all emails to your Interac e-Transfer
            contacts. If incorrect,
            {' '}
            <a href="/" className="text-black">
              edit your profile.
            </a>
          </p>
        </Col>
      </Row>
      <Row className="mt-2">
        <Col md={4}>
          <Form.Label>Your Email Nickname:</Form.Label>
        </Col>
        <Col md={8}>
          <Form.Label>{getUserFullName()}</Form.Label>
        </Col>
      </Row>
      <Row className="mt-2">
        <Col md={4}>
          <Form.Label>Your Email Address:</Form.Label>
        </Col>
        <Col md={8}>
          <Form.Label>{getUserEmail()}</Form.Label>
        </Col>
      </Row>
      <hr style={{ height: '1px', width: '100%' }} />
      <Row className="mt-4 mb-4">
        <Col>
          <Button variant="light" className="d-flex">
            <div
              style={{
                width: 20,
                height: 25,
                border: '1px dotted grey',
                textAlign: 'center',
                marginRight: 10,
              }}
            >
              P
            </div>
            Cancel
          </Button>
        </Col>
        <Col className="d-flex justify-content-end">
          <Button
            className="zippy-btn d-flex"
            onClick={handleSecurity}
            disabled={isProcessing}
          >
            {isProcessing ? (
              <span>
                Next
                <Spinner style={{ marginLeft: '16px' }} animation="border" variant="light" size="sm" />
              </span>
            ) : (
              'Next'
            )}
          </Button>
        </Col>
      </Row>
    </>
  );
};

export default DetailsPage;
