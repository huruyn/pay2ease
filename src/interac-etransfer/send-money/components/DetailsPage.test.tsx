import
{
  cleanup,
  fireEvent,
  render,
  screen,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import DetailsPage from './DetailsPage';
import MAIN_INFO from '../../../stories/MainInfo';

const mockSetMainInfo = jest.fn();
const mockHandleSecurity = jest.fn();
const mockSetContactToSend = jest.fn();

const component = (
  <DetailsPage
    setMainInfo={mockSetMainInfo}
    mainInfo={MAIN_INFO}
    setContactToSend={mockSetContactToSend}
    setSelectedAccount={jest.fn()}
    selectedAccount={1}
    selectedContact={1}
    contacts={[
      {
        id: 1,
        firstName: 'Test',
        lastName: 'Test',
        email: 'email@zippy.cash',
        phone: 'phone',
      },
    ]}
    accounts={[
      {
        id: 0,
        name: 'test',
        balance: 0,
      },
    ]}
    handleSecurity={mockHandleSecurity}
    isProcessing={false}
    user={undefined}
  />
);

beforeEach(cleanup);

describe('DetailsPage Component', () => {
  it('should render SecurityQuestionPage', () => {
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });

  it('Click next button on DetailsPage', () => {
    render(component);
    screen.getAllByRole('button')[1].click();
    expect(mockHandleSecurity).toBeCalledTimes(1);
  });

  it('Click next button on DetailsPage with option changing', () => {
    render(component);
    userEvent.selectOptions(screen.getAllByRole('combobox')[0], ['1']);
    screen.getAllByRole('button')[1].click();
    expect(mockHandleSecurity).toBeCalledTimes(1);
    expect(mockSetContactToSend).toBeCalledTimes(1);
  });

  it('Click next button on DetailsPage with option 2', () => {
    render(component);
    screen.getAllByRole('button')[1].click();
    expect(mockHandleSecurity).toBeCalledTimes(1);
  });

  it('change FormControl text values', () => {
    render(component);

    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: 'new message' },
    });
    fireEvent.change(screen.getByDisplayValue('1000'), {
      target: { value: 200 },
    });

    expect(mockSetMainInfo).toBeCalledTimes(2);
  });

  it('change FormControl to call setContactToSend', () => {
    render(component);
    userEvent.selectOptions(screen.getAllByRole('combobox')[0], ['1']);
    expect(mockSetContactToSend).toBeCalledTimes(1);
  });

  it('next button click should not affect ammount', () => {
    render(component);
    expect(screen.getByDisplayValue('1000')).toBeInTheDocument();
    screen.getAllByRole('button')[1].click();
    expect(screen.getByDisplayValue('1000')).toBeInTheDocument();
  });

  it('next button click should not affect transfer method', () => {
    render(component);
    expect(screen.getAllByRole('combobox')[2].textContent).toEqual('Email');
    screen.getAllByRole('button')[1].click();
    expect(screen.getAllByRole('combobox')[2].textContent).toEqual('Email');
  });

  it('getEmail without contacts should render no email', () => {
    const wrapper = (
      <DetailsPage
        setMainInfo={mockSetMainInfo}
        mainInfo={MAIN_INFO}
        setContactToSend={mockSetContactToSend}
        setSelectedAccount={jest.fn()}
        selectedAccount={0}
        contacts={[]}
        accounts={[
          {
            id: 0,
            name: 'test',
            balance: 0,
          },
        ]}
        selectedContact={1}
        handleSecurity={jest.fn()}
        isProcessing={false}
        user={undefined}
      />
    );

    render(wrapper);
    expect(wrapper).toMatchSnapshot();
  });

  it('getEmail with contacts should render an email', () => {
    render(component);
    expect(component).toMatchSnapshot();
  });

  it('should not call setCurrentStep when no contact selected', () => {
    const wrapper = (
      <DetailsPage
        setMainInfo={mockSetMainInfo}
        mainInfo={MAIN_INFO}
        setContactToSend={mockSetContactToSend}
        setSelectedAccount={jest.fn()}
        selectedAccount={1}
        selectedContact={0}
        contacts={[
          {
            id: 1,
            firstName: 'Test',
            lastName: 'Test',
            email: 'email@zippy.cash',
            phone: 'phone',
          },
        ]}
        accounts={[
          {
            id: 0,
            name: 'test',
            balance: 0,
          },
        ]}
        handleSecurity={jest.fn()}
        isProcessing={false}
        user={undefined}
      />
    );

    render(wrapper);
    expect(mockHandleSecurity).toBeCalledTimes(0);
  });

  it('should not call setCurrentStep when no account selected', () => {
    const wrapper = (
      <DetailsPage
        setMainInfo={mockSetMainInfo}
        mainInfo={MAIN_INFO}
        setContactToSend={mockSetContactToSend}
        setSelectedAccount={jest.fn()}
        selectedAccount={0}
        selectedContact={1}
        contacts={[
          {
            id: 1,
            firstName: 'Test',
            lastName: 'Test',
            email: 'email@zippy.cash',
            phone: 'phone',
          },
        ]}
        accounts={[
          {
            id: 0,
            name: 'test',
            balance: 0,
          },
        ]}
        handleSecurity={jest.fn()}
        isProcessing={false}
        user={undefined}
      />
    );

    render(wrapper);
    expect(mockHandleSecurity).toBeCalledTimes(0);
  });

  it('should not call setCurrentStep when no transferMethod selected', () => {
    const wrapper = (
      <DetailsPage
        setMainInfo={mockSetMainInfo}
        mainInfo={{ ...MAIN_INFO, transferMethod: '' }}
        setContactToSend={mockSetContactToSend}
        setSelectedAccount={jest.fn()}
        selectedAccount={1}
        selectedContact={1}
        contacts={[
          {
            id: 1,
            firstName: 'Test',
            lastName: 'Test',
            email: 'email@zippy.cash',
            phone: 'phone',
          },
        ]}
        accounts={[
          {
            id: 0,
            name: 'test',
            balance: 0,
          },
        ]}
        handleSecurity={jest.fn()}
        isProcessing={false}
        user={undefined}
      />
    );

    render(wrapper);
    expect(mockHandleSecurity).toBeCalledTimes(0);
  });

  it('should not call setCurrentStep when ammount equal 0', () => {
    const wrapper = (
      <DetailsPage
        setMainInfo={mockSetMainInfo}
        mainInfo={{ ...MAIN_INFO, amount: 0 }}
        setContactToSend={mockSetContactToSend}
        setSelectedAccount={jest.fn()}
        selectedAccount={1}
        selectedContact={1}
        contacts={[
          {
            id: 1,
            firstName: 'Test',
            lastName: 'Test',
            email: 'email@zippy.cash',
            phone: 'phone',
          },
        ]}
        accounts={[
          {
            id: 0,
            name: 'test',
            balance: 0,
          },
        ]}
        handleSecurity={jest.fn()}
        isProcessing={false}
        user={undefined}
      />
    );

    render(wrapper);
    expect(mockHandleSecurity).toBeCalledTimes(0);
  });

  it('should not call setCurrentStep when ammount greater than 3000', () => {
    const wrapper = (
      <DetailsPage
        setMainInfo={mockSetMainInfo}
        mainInfo={{ ...MAIN_INFO, amount: 3001 }}
        setContactToSend={mockSetContactToSend}
        setSelectedAccount={jest.fn()}
        selectedAccount={1}
        selectedContact={1}
        contacts={[
          {
            id: 1,
            firstName: 'Test',
            lastName: 'Test',
            email: 'email@zippy.cash',
            phone: 'phone',
          },
        ]}
        accounts={[
          {
            id: 0,
            name: 'test',
            balance: 0,
          },
        ]}
        handleSecurity={jest.fn()}
        isProcessing={false}
        user={undefined}
      />
    );

    render(wrapper);
    expect(mockHandleSecurity).toBeCalledTimes(0);
  });
});
