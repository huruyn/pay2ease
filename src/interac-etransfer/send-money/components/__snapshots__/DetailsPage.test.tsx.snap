// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsPage Component getEmail with contacts should render an email 1`] = `
<DetailsPage
  accounts={
    Array [
      Object {
        "balance": 0,
        "id": 0,
        "name": "test",
      },
    ]
  }
  contacts={
    Array [
      Object {
        "email": "email@zippy.cash",
        "firstName": "Test",
        "id": 1,
        "lastName": "Test",
        "phone": "phone",
      },
    ]
  }
  handleSecurity={[MockFunction]}
  isProcessing={false}
  mainInfo={
    Object {
      "amount": 1000,
      "confirmSecurityAnswer": "security Answer",
      "destination": Object {
        "email": "destination@email.com",
        "name": "destination name",
      },
      "fromAccount": "fromAccount (8000 0000 00000000)",
      "message": "message",
      "securityAnswer": "security Answer",
      "securityQuestion": "security Question",
      "showAnswer": true,
      "source": Object {
        "email": "source@email.com",
        "name": "source name",
      },
      "transferMethod": "Email",
    }
  }
  selectedAccount={1}
  selectedContact={1}
  setContactToSend={[MockFunction]}
  setMainInfo={[MockFunction]}
  setSelectedAccount={[MockFunction]}
/>
`;

exports[`DetailsPage Component getEmail without contacts should render no email 1`] = `
<DetailsPage
  accounts={
    Array [
      Object {
        "balance": 0,
        "id": 0,
        "name": "test",
      },
    ]
  }
  contacts={Array []}
  handleSecurity={[MockFunction]}
  isProcessing={false}
  mainInfo={
    Object {
      "amount": 1000,
      "confirmSecurityAnswer": "security Answer",
      "destination": Object {
        "email": "destination@email.com",
        "name": "destination name",
      },
      "fromAccount": "fromAccount (8000 0000 00000000)",
      "message": "message",
      "securityAnswer": "security Answer",
      "securityQuestion": "security Question",
      "showAnswer": true,
      "source": Object {
        "email": "source@email.com",
        "name": "source name",
      },
      "transferMethod": "Email",
    }
  }
  selectedAccount={0}
  selectedContact={1}
  setContactToSend={[MockFunction]}
  setMainInfo={[MockFunction]}
  setSelectedAccount={[MockFunction]}
/>
`;

exports[`DetailsPage Component should render SecurityQuestionPage 1`] = `
<div>
  <h4>
    Your Interac e-Transfer Details
  </h4>
  <ul>
    <li>
      Fees apply to Interac e-Transfer transactions.
       
      <a
        class="text-black"
        href="/"
      >
        Learn more about Interac e-Transfer fees
      </a>
    </li>
    <li>
      There are limits on how much money you can send in a day, a week and a month.
       
      <a
        class="text-black"
        href="/"
      >
        Learn more about Interac e-Transfer limits
      </a>
    </li>
    <li>
      Note: Sending money outside of Canada? Use the
       
      <a
        class="text-black"
        href="/"
      >
        ZippyX Global Money Transfer Services
      </a>
    </li>
  </ul>
  <div
    class="mt-4 row"
  >
    <div
      class="col-md-4"
    >
      <label
        class="form-label"
      >
        Send Money To:
      </label>
    </div>
    <div
      class="col-md-8"
    >
      <select
        class="send-account-select form-select"
      >
        <option>
          Select
        </option>
        <option
          value="1"
        >
          Test Test
        </option>
      </select>
      <div
        class="contact-details"
      >
        <div>
          Email: email@zippy.cash
        </div>
      </div>
    </div>
  </div>
  <div
    class="mt-2 row"
  >
    <div
      class="col-md-4"
    >
      <label
        class="form-label"
      >
        Amount:
      </label>
    </div>
    <div
      class="col-md-8"
    >
      <input
        class="form-control"
        step=".01"
        type="number"
        value="1000"
      />
      <label
        class="form-label"
      >
        The maximum amount you can send in each transfer is $3,000.
      </label>
    </div>
  </div>
  <div
    class="mt-2 row"
  >
    <div
      class="col-md-4"
    >
      <label
        class="form-label"
      >
        From Account:
      </label>
    </div>
    <div
      class="col-md-8"
    >
      <select
        class="from-account-info form-select"
      >
        <option>
          Select
        </option>
        <option
          value="0"
        >
          test
        </option>
      </select>
    </div>
  </div>
  <div
    class="mt-2 row"
  >
    <div
      class="col-md-4"
    >
      <label
        class="form-label"
      >
        Transfer Method:
      </label>
    </div>
    <div
      class="col-md-8"
    >
      <select
        class="transfer-method form-select"
      >
        <option>
          Email
        </option>
      </select>
    </div>
  </div>
  <div
    class="mt-2 row"
  >
    <div
      class="col-md-4"
    >
      <label
        class="form-label"
      >
        Message (Optional):
      </label>
    </div>
    <div
      class="col-md-8"
    >
      <textarea
        class="form-control"
        rows="3"
      >
        message
      </textarea>
      <p>
        Do not provide the security question, any part of the security answer or any confidential information in your message to the contact. This message will be viewable on the Interac e-Transfer Status page.
      </p>
    </div>
  </div>
  <hr
    style="height: 1px; width: 100%;"
  />
  <div
    class="mt-2 row"
  >
    <div
      class="col-md-12"
    >
      <p>
        This is how you will appear in all emails to your Interac e-Transfer contacts. If incorrect,
         
        <a
          class="text-black"
          href="/"
        >
          edit your profile.
        </a>
      </p>
    </div>
  </div>
  <div
    class="mt-2 row"
  >
    <div
      class="col-md-4"
    >
      <label
        class="form-label"
      >
        Your Email Nickname:
      </label>
    </div>
    <div
      class="col-md-8"
    >
      <label
        class="form-label"
      />
    </div>
  </div>
  <div
    class="mt-2 row"
  >
    <div
      class="col-md-4"
    >
      <label
        class="form-label"
      >
        Your Email Address:
      </label>
    </div>
    <div
      class="col-md-8"
    >
      <label
        class="form-label"
      />
    </div>
  </div>
  <hr
    style="height: 1px; width: 100%;"
  />
  <div
    class="mt-4 mb-4 row"
  >
    <div
      class="col"
    >
      <button
        class="d-flex btn btn-light"
        type="button"
      >
        <div
          style="width: 20px; height: 25px; border: 1px dotted grey; text-align: center; margin-right: 10px;"
        >
          P
        </div>
        Cancel
      </button>
    </div>
    <div
      class="d-flex justify-content-end col"
    >
      <button
        class="zippy-btn d-flex btn btn-primary"
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
